src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs

```diff
@@ -5,16 +5,81 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Reflection;
 using System.Text;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
 
 namespace Microsoft.Interop
 {
     public static class IncrementalGeneratorInitializationContextExtensions
     {
+        // This type is a record to get the generated equality and hashing operators
+        // which will be faster than the reflection-based ones.
+        private record struct TargetFrameworkSettings
+        {
+            public TargetFrameworkSettings(AnalyzerConfigOptions options)
+            {
+                options.TryGetValue("build_property.TargetFrameworkIdentifier", out string? frameworkIdentifier);
+                Identifier = frameworkIdentifier;
+                options.TryGetValue("build_property.TargetFrameworkVersion", out string? version);
+                // TargetFrameworkVersion starts with a 'v'.
+                Version = version is not null ? Version.Parse(version.Substring(1)) : null;
+            }
+
+            public string? Identifier { get; init; }
+
+            public Version? Version { get; init; }
+        }
+
+        private static readonly Version FirstNonCoreVersion = new(5, 0);
+
+        // Parse from the informational version as that is the only version that always matches the TFM version
+        // even in debug builds.
+        private static readonly Version ThisAssemblyVersion = Version.Parse(
+            typeof(IncrementalGeneratorInitializationContextExtensions).Assembly
+            .GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion.Split('-')[0]);
```

Comment:
Doing a throwing parse in a static initializer always makes me a bit nervous. But if you all have test cases around this i'll feel a bit better.

Reply:
The majority of our test suite is going down the route that will use this member, so it's heavily covered in our testing.